bin/public_interface.py

```diff
@@ -91,21 +95,133 @@ def _print(signature: Dict[str, inspect.Signature], variables: Set[str]) -> None
+            continue
+        if (
+            original_argument["name"] == arguments[i]["name"]
+            and original_argument["kind"] == arguments[i]["kind"]
+            and "default" not in original_argument
+            and "default" in arguments[i]
+        ):
+            continue
+        return False
+    # it is an optional argument when it has a default value:
+    return all(["default" in argument for argument in arguments[len(original_arguments) :]])
+
+
+def _is_compatible(original_arguments: List[Dict[str, Any]], arguments: List[Dict[str, Any]]) -> bool:
+    """
+    If there is an argument change, it is compatible only when
+    - new optional arguments are added or existing arguments become optional.
+    - self is removed (method -> staticmethod).
+    - both of them combined
+    """
+    if original_arguments == arguments or _only_new_optional_arguments_or_existing_arguments_optionalized(
+        original_arguments, arguments
+    ):
+        return True
+    if original_arguments and original_arguments[0] == _ARGUMENT_SELF:
+        original_arguments_without_self = original_arguments[1:]
+        if _is_compatible(original_arguments_without_self, arguments):
+            return True
+    return False
+
+
+def _detect_breaking_changes(
+    original_routines: Dict[str, List[Dict[str, Any]]],
+    original_variables: Set[str],
+    routines: Dict[str, List[Dict[str, Any]]],
+    variables: Set[str],
+) -> _BreakingChanges:
+    deleted_routines: List[str] = []
+    incompatible_routines: List[str] = []
+    for routine_path, arguments in original_routines.items():
+        if routine_path not in routines:
+            deleted_routines.append(routine_path)
+        elif not _is_compatible(arguments, routines[routine_path]):
+            incompatible_routines.append(routine_path)
+    return _BreakingChanges(
+        sorted(set(original_variables) - set(variables)), sorted(deleted_routines), sorted(incompatible_routines)
+    )
+
+
 def main() -> None:
     parser = argparse.ArgumentParser(description=__doc__)
 
     subparsers = parser.add_subparsers(dest="command")
     extract = subparsers.add_parser("extract", help="Extract public interfaces")
     extract.add_argument("--module", help="The module to extract public interfaces", type=str, default="samtranslator")
+    check = subparsers.add_parser("check", help="Check public interface changes")
+    check.add_argument("original_json", help="The original public interface JSON file", type=Path)
+    check.add_argument("new_json", help="The new public interface JSON file", type=Path)
     args = parser.parse_args()
 
     if args.command == "extract":
         scanner = InterfaceScanner()
         scanner.scan_interfaces_recursively(args.module)
         _print(scanner.signatures, scanner.variables)
-    # TODO: handle compare
+    elif args.command == "check":
+        with open(args.original_json, "r") as original_f, open(args.new_json, "r") as new_f:
```

Comment:
A cleaner approach is get rid of the `with`, and do `new_json = json.loads(args.new_json.read_text())` directly.

Reply:
let me try